working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+> object graphs.
+>
+> It is worth highlighting **shared memory multithreading does not necessarily
+> imply simultaneous access to mutable data.** Developers can still structure
+> their parallel code using isolates and message passing - but they can avoid
+> the cost of copying the data by sending the message which can be directly
+> shared with the receiver rather than copied.
+
+[`Isolate.run`]: https://api.dart.dev/stable/3.2.6/dart-isolate/Isolate/run.html
+
+### Interoperability
+
+Consider the following C code using a miniaudio library:
+
+```cpp
+static void DataCallback(
+   ma_device* device, void* out, const void* in, ma_uint32 frame_count) {
+  // Synchronously process up to |frame_count| frames.
+}
+
+ma_device_config config = ma_device_config_init(ma_device_type_playback);
+// This function will be called when miniaudio needs more data.
+// The call will happend on a backend specific thread dedicated to
+// audio playback.
+config.dataCallback      = &DataCallback;
+// ...
+```
+
+Porting this code to Dart using `dart:ffi` is currently impossible, as FFI only
+supports two specific callback types:
+
+- [`NativeCallable.isolateLocal`][native-callable-isolate-local]: native caller
+  must have an exclusive access to an isolate in which callback was created.
+  This type of callback works if Dart calls C and C calls back into Dart
+  synchronously. It also works if caller uses VM C API for entering isolates
+  (e.g.`Dart_EnterIsolate`/`Dart_ExitIsolate`).
+- [`NativeCallable.listener`][native-callable-listener]: native caller
+  effectively sends a message to the isolate which created the callback and does
+  not synchronously wait for the response.
+
+Neither of these work for this use-case where native caller wants to perform a
+synchronous invocation. There are obvious ways to address this:
+
+1. Create a variation of `NativeCallable.isolateLocal` which enters (and after
+   call leaves) the target isolate if the native caller is not already in the
+   target isolate.
+
+2. Create `NativeCallable.onTemporaryIsolate` which spawns (and after call
+   destroys) a temporary isolate to handle the call.
+
+Neither of these are truly satisfactory:
+
+- Allowing `isolateLocal` to enter target isolate means that it can block the
+  caller if the target isolate is busy doing _something_ (e.g. processing some
+  events) in parallel. This is unacceptable in situations when the caller is
+  latency sensitive e.g. audio thread or even just the main UI thread of an
+  application.
+- Using temporary isolate comes with a bunch of ergonomic problems as every
+  invocation is handled using a freshly created environment and no static state
+  is carried between invocations - which might surprise the developer. Sharing
+  mutable data between invocations requires storing it outside of Dart heap
+  using `dart:ffi`.
+
+> [!NOTE]
+>
+> This particular example might not look entirely convincing because it can be
```

Comment:
I don't understand this sentence - you just convinced me that there is no reasonable solution using the isolate model!